src/mpl2/src/clusterEngine.cpp

```diff
@@ -0,0 +1,2160 @@
+    for (auto& inst : cluster->getLeafMacros()) {
+      tree_->maps.inst_to_cluster_id[inst] = cluster_id;
+    }
+  }
+
+  if (cluster_type == StdCellCluster || cluster_type == MixedCluster) {
+    for (auto& inst : cluster->getLeafStdCells()) {
+      tree_->maps.inst_to_cluster_id[inst] = cluster_id;
+    }
+  }
+
+  // Note: macro clusters have no module.
+  if (cluster_type == StdCellCluster) {
+    for (auto& module : cluster->getDbModules()) {
+      updateInstancesAssociation(module, cluster_id, false);
+    }
+  } else if (cluster_type == MixedCluster) {
+    for (auto& module : cluster->getDbModules()) {
+      updateInstancesAssociation(module, cluster_id, true);
+    }
+  }
+}
+
+// Unlike macros, std cells are always considered when when updating
+// the inst -> cluster map with the data from a module.
+void ClusteringEngine::updateInstancesAssociation(odb::dbModule* module,
+                                                  int cluster_id,
+                                                  bool include_macro)
+{
+  if (include_macro) {
+    for (odb::dbInst* inst : module->getInsts()) {
+      tree_->maps.inst_to_cluster_id[inst] = cluster_id;
+    }
+  } else {  // only consider standard cells
+    for (odb::dbInst* inst : module->getInsts()) {
+      odb::dbMaster* master = inst->getMaster();
+      if (isIgnoredMaster(master) || master->isBlock()) {
+        continue;
+      }
+
+      tree_->maps.inst_to_cluster_id[inst] = cluster_id;
+    }
+  }
+  for (odb::dbModInst* inst : module->getChildren()) {
+    updateInstancesAssociation(inst->getMaster(), cluster_id, include_macro);
+  }
+}
+
+void ClusteringEngine::setClusterMetrics(Cluster* cluster)
+{
+  float std_cell_area = 0.0f;
+  for (odb::dbInst* std_cell : cluster->getLeafStdCells()) {
+    std_cell_area += computeMicronArea(std_cell);
+  }
+
+  float macro_area = 0.0f;
+  for (odb::dbInst* macro : cluster->getLeafMacros()) {
+    macro_area += computeMicronArea(macro);
+  }
+
+  const unsigned int num_std_cell = cluster->getLeafStdCells().size();
+  const unsigned int num_macro = cluster->getLeafMacros().size();
+
+  Metrics metrics(num_std_cell, num_macro, std_cell_area, macro_area);
+
+  for (auto& module : cluster->getDbModules()) {
+    metrics.addMetrics(*tree_->maps.module_to_metrics[module]);
+  }
+
+  debugPrint(logger_,
+             MPL,
+             "multilevel_autoclustering",
+             1,
+             "Setting Cluster Metrics for {}: Num Macros: {} Num Std Cells: {}",
+             cluster->getName(),
+             metrics.getNumMacro(),
+             metrics.getNumStdCell());
+
+  if (cluster->getClusterType() == HardMacroCluster) {
+    cluster->setMetrics(
+        Metrics(0, metrics.getNumMacro(), 0.0, metrics.getMacroArea()));
+  } else if (cluster->getClusterType() == StdCellCluster) {
+    cluster->setMetrics(
+        Metrics(metrics.getNumStdCell(), 0, metrics.getStdCellArea(), 0.0));
+  } else {
+    cluster->setMetrics(metrics);
+  }
```

Comment:
The cluster knows its type so the caller should only need to do `cluster->setMetrics(metrics)` and the setMetrics method should handle the rest.

Reply:
These checks are not even needed. The module Metrics is populated according to its instances so if a module has e.g., no std cells its Metrics will also not have std cells and it's not needed to create a new Metrics for HardMacroCluster as the Metrics created earlier in the function will have the correct data.